copy/sign.go

```diff
@@ -29,3 +40,53 @@ func (c *copier) createSignature(manifest []byte, keyIdentity string) ([]byte, e
 	}
 	return newSig, nil
 }
+
+// createCosignSignature creates a new signature of manifest.
+func (c *copier) createCosignSignature(manifest []byte) ([]byte, error) {
+	mech, err := signature.NewSigstoreSigningMechanism()
+	if err != nil {
+		return nil, errors.Wrap(err, "initializing Sigstore")
+	}
+	defer mech.Close()
+	if err := mech.SupportsSigning(); err != nil {
+		return nil, errors.Wrap(err, "signing is not supported")
+	}
+
+	dockerReference := c.dest.Reference().DockerReference()
+	if dockerReference == nil {
+		return nil, errors.Errorf("Cannot determine canonical Docker reference for destination %s", transports.ImageName(c.dest.Reference()))
+	}
+
+	manifestDigest, err := man.Digest(manifest)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println("Generating certificate from Fulcio")
+	if err := mech.GenerateCertificate(); err != nil {
+		return nil, errors.Wrap(err, "getting key from Fulcio")
+	}
+
+	fmt.Println("Marshalling payload into JSON")
+	cosignSignature := signature.NewCosignSignature(manifestDigest, dockerReference.String())
+	sigPayload, err := cosignSignature.MarshalJSON()
```

Comment:
This feels like something that should be a bit higher-level; at this high level of abstraction, the caller shouldn’t need to do 4 calls exactly in sequence, and definitely not care that the format is JSON. Maybe, very vaguely

- SignDockerManifestWithCosign(manifest, imageIdentity, whoSignsThis) -> blob(s)ToStoreWithTheImage, dataForUpload
- UploadTransparencyLogEntry(dataForUpload, credentialsIfAny)

(Should the mechanism even have _any_ public methods apart from `Close`? The `Sign…WithCosign` might be the only supported caller for now. We can always make private methods public, but not the other way.)